numpy/numpy
# ENH, DOC: Add support for interactive examples for NumPy with `jupyterlite-sphinx`

## Description

This PR adds support for WASM-powered interactive examples with [JupyterLite](https://jupyterlite.readthedocs.io/en/stable/) and [`jupyterlite-sphinx`](https://jupyterlite-sphinx.readthedocs.io/) in the NumPy documentation.

A similar enhancement was made for the documentation for SciPy (https://github.com/scipy/scipy/pull/20019) and that for PyWavelets (https://github.com/PyWavelets/pywt/pull/728).

### How to test this

> [!TIP]
> I have a GitHub Pages website deployed at https://agriyakhetarpal.github.io/numpy/ through which these changes can be experimented with. The CircleCI deployment does not render the examples properly, however, a fix is available as reported in https://github.com/numpy/numpy/pull/26745#issuecomment-2183812333 and will be fixed upstream in https://github.com/jupyterlite/jupyterlite-sphinx/pull/182.
>
> **Update, 24/07/2024:** this fix is now available in `jupyterlite-sphinx` version 0.16.2, and the CircleCI deployment should render the examples properly as well.

### Key changes (so far)

1. Configuration for `jupyterlite-sphinx` in `doc/source/conf.py` and `doc/Makefile`, coupled with changes to styling in `doc/source/_static/numpy.css`
2. Updates to the API documentation under files in `numpy/` under the "Examples" section in the docstrings for various modules

> [!NOTE]
> I have currently left out the following: `_datasource.py`, `_utils_impl.py`, `_version.py`, `_array_api_info.py`, `ctypeslib.py`, files under `numpy/_typing/`, and `numpy/_core/numerictypes.py`.

> [!TIP]
> The "NumPy user guide" section has not been modified at this time due to the size of the PR (in terms of the number of files modified). However, it can be modified or adapted to a Jupytext notebook similar to scipy/scipy#20303 to make it work well in an interactive set-up.

> [!IMPORTANT]
> The NumPy documentation will have to be rebuilt and reuploaded in https://github.com/numpy/doc/ for these changes to show up in the official documentation.

## What issue does this PR address?

Closes #25969

## Additional context

- Updates to NumPy in-tree are coupled with new releases for Pyodide, and Pyodide 0.27 as of 01/01/2025 brings NumPy version 2.0.2, which is slightly behind NumPy's latest release at the time of writing. Therefore, some examples or new features may not work as expected and can fail in some cases, for which a global warning text has been provided and displayed before each interactive example, and examples that are known to fail have been disabled via [the `try_examples.json` configuration file](https://jupyterlite-sphinx.readthedocs.io/en/stable/directives/try_examples.html#try-examples-json-configuration-file). A custom Pyodide distribution or a custom wheel of NumPy is not provided with this PR, and a solution is being explored with nightly WASM wheels for NumPy (established in gh-26134) via https://github.com/pyodide/pyodide/issues/4898.
- SciPy documentation: https://docs.scipy.org/doc/scipy/reference/index.html
- PyWavelets documentation: https://pywavelets.readthedocs.io/en/latest/ref/index.html
- xref: https://github.com/Quansight-Labs/czi-scientific-python-mgmt/issues/19
